x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSenderTests.java

```diff
@@ -276,14 +278,78 @@ public void testNonFatalFailedOnAllNodes() {
         assertThat(sent.size(), equalTo(1));
     }
 
+    public void testLimitConcurrentNodes() {
+        var targetShards = List.of(
+            targetShard(shard1, node1),
+            targetShard(shard2, node2),
+            targetShard(shard3, node3),
+            targetShard(shard4, node4),
+            targetShard(shard5, node5)
+        );
+
+        var concurrency = randomIntBetween(1, targetShards.size() - 1);
+        AtomicInteger maxConcurrentRequests = new AtomicInteger(0);
+        AtomicInteger concurrentRequests = new AtomicInteger(0);
+        Queue<NodeRequest> sent = ConcurrentCollections.newQueue();
+        var response = safeGet(sendRequests(targetShards, randomBoolean(), concurrency, (node, shardIds, aliasFilters, listener) -> {
+            concurrentRequests.incrementAndGet();
+
+            while (true) {
+                var priorMax = maxConcurrentRequests.get();
+                var newMax = Math.max(priorMax, concurrentRequests.get());
+                if (newMax <= priorMax || maxConcurrentRequests.compareAndSet(priorMax, newMax)) {
+                    break;
+                }
+            }
+
+            sent.add(new NodeRequest(node, shardIds, aliasFilters));
+            runWithDelay(() -> {
+                concurrentRequests.decrementAndGet();
+                listener.onResponse(new DataNodeComputeResponse(List.of(), Map.of()));
+            });
+        }));
+        assertThat(sent.size(), equalTo(5));
+        assertThat(maxConcurrentRequests.get(), equalTo(concurrency));
+        assertThat(response.totalShards, equalTo(5));
+        assertThat(response.successfulShards, equalTo(5));
+        assertThat(response.failedShards, equalTo(0));
+    }
+
+    public void testDoesNotSendMoreRequestsAfterNodeIsSkipped() {
+        var targetShards = List.of(
+            targetShard(shard1, node1),
+            targetShard(shard2, node2),
+            targetShard(shard3, node3),
+            targetShard(shard4, node4),
+            targetShard(shard5, node5)
+        );
+
+        AtomicInteger processed = new AtomicInteger(0);
+        Queue<NodeRequest> sent = ConcurrentCollections.newQueue();
+        var response = safeGet(sendRequests(targetShards, randomBoolean(), 1, (node, shardIds, aliasFilters, listener) -> {
+            sent.add(new NodeRequest(node, shardIds, aliasFilters));
+            runWithDelay(() -> {
+                if (processed.incrementAndGet() == 1) {
+                    listener.onResponse(new DataNodeComputeResponse(List.of(), Map.of()));
+                } else {
+                    listener.onSkip(true);
+                }
+            });
+        }));
+        assertThat(sent.size(), equalTo(2));// onResponse() + onSkip()
+        assertThat(response.totalShards, equalTo(5));
+        assertThat(response.successfulShards, equalTo(5));
+        assertThat(response.failedShards, equalTo(0));
```

Comment:
I wonder if we should count skipped shards as successful or as `org.elasticsearch.xpack.esql.plugin.ComputeResponse#skippedShards`